fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -292,19 +313,33 @@ struct RocksDBKeyValueStore : IKeyValueStore {
 		struct ReadRangeAction : TypedAction<Reader, ReadRangeAction>, FastAllocated<ReadRangeAction> {
 			KeyRange keys;
 			int rowLimit, byteLimit;
+			double startTime;
 			ThreadReturnPromise<RangeResult> result;
 			ReadRangeAction(KeyRange keys, int rowLimit, int byteLimit)
-			  : keys(keys), rowLimit(rowLimit), byteLimit(byteLimit) {}
+			  : keys(keys), rowLimit(rowLimit), byteLimit(byteLimit), startTime(timer_monotonic()) {}
 			double getTimeEstimate() const override { return SERVER_KNOBS->READ_RANGE_TIME_ESTIMATE; }
 		};
 		void action(ReadRangeAction& a) {
+			if (timer_monotonic() - a.startTime > SERVER_KNOBS->ROCKSDB_READ_RANGE_TIMEOUT) {
+				TraceEvent(SevWarn, "RocksDBError")
+				    .detail("Error", "Read range request timedout")
+				    .detail("Method", "ReadRangeAction")
+				    .detail("Timeout value", SERVER_KNOBS->ROCKSDB_READ_RANGE_TIMEOUT);
+				a.result.sendError(transaction_too_old());
+				return;
+			}
+
 			RangeResult result;
 			if (a.rowLimit == 0 || a.byteLimit == 0) {
 				a.result.send(result);
 			}
 			int accumulatedBytes = 0;
 			rocksdb::Status s;
 			auto options = getReadOptions();
+			std::chrono::seconds readRangeTimeout(SERVER_KNOBS->ROCKSDB_READ_RANGE_TIMEOUT);
+			options.deadline = std::chrono::duration_cast<std::chrono::microseconds>(
+			                       std::chrono::system_clock::now().time_since_epoch()) +
```

Comment:
From the documentation (https://github.com/facebook/rocksdb/blob/f4529a54bba97ca81c83e74c87cf780377096f01/include/rocksdb/options.h#L1496), it seems that this should use `db->GetEnv()->NowMicros()` to ensure we don't have issues with clock skew.